libs/pinecone/langchain_pinecone/_utilities.py

```diff
@@ -185,3 +189,117 @@ def sparse_cosine_similarity(X: SparseValues, Y: List[SparseValues]) -> np.ndarr
+            allowed_vector_types = ("dense", "sparse", None)
+            if model_type not in allowed_model_types:
+                raise ValueError(
+                    f"model_type must be one of {allowed_model_types}, got {model_type}"
+                )
+            if vector_type not in allowed_vector_types:
+                raise ValueError(
+                    f"vector_type must be one of {allowed_vector_types}, got {vector_type}"
+                )
+            return model_type, vector_type
+
+    # Validate arguments
+    _ModelParamsModel.validate(model_type, vector_type)
+
+    pc = Pinecone(api_key=api_key)
+    kwargs = {}
+    if model_type:
+        kwargs["type"] = model_type
+    if vector_type:
+        kwargs["vector_type"] = vector_type
+    return pc.inference.list_models(**kwargs)
+
+
+async def aget_pinecone_supported_models(
+    api_key: str, model_type: Optional[str] = None, vector_type: Optional[str] = None
+) -> list:
+    """Fetch supported models from Pinecone dynamically using async HTTP calls.
+    Args:
+        api_key: Pinecone API key
+        model_type: 'embed', 'rerank', or None for all
+        vector_type: 'dense', 'sparse', or None
+    Returns:
+        List of model info dicts
+    Raises:
+        ValueError: if model_type or vector_type is not allowed
+        httpx.HTTPError: if the API request fails
+    """
+
+    class _ModelParamsModel:
+        model_type: Optional[Literal["embed", "rerank"]] = None
+        vector_type: Optional[Literal["dense", "sparse"]] = None
+
+        @classmethod
+        def validate(
+            cls, model_type: Optional[str], vector_type: Optional[str]
+        ) -> tuple[Optional[str], Optional[str]]:
+            # Pydantic-style validation
+            allowed_model_types = ("embed", "rerank", None)
+            allowed_vector_types = ("dense", "sparse", None)
+            if model_type not in allowed_model_types:
+                raise ValueError(
+                    f"model_type must be one of {allowed_model_types}, got {model_type}"
+                )
+            if vector_type not in allowed_vector_types:
+                raise ValueError(
+                    f"vector_type must be one of {allowed_vector_types}, got {vector_type}"
+                )
+            return model_type, vector_type
+
+    # Validate arguments
+    _ModelParamsModel.validate(model_type, vector_type)
+
+    # Pinecone API base URL
+    base_url = "https://api.pinecone.io"
+    endpoint = "/models"
+
+    # Build query parameters
+    params = {}
+    if model_type:
+        params["type"] = model_type
+    if vector_type:
+        params["vector_type"] = vector_type
+
+    # Headers for authentication
+    headers = {
+        "Api-Key": api_key,
+        "Content-Type": "application/json",
+        "X-Pinecone-API-Version": "2025-04",
+    }
+
+    async with httpx.AsyncClient() as client:
+        response = await client.get(
+            f"{base_url}{endpoint}",
```

Comment:
```suggestion
            f"{base_url}/models",
```

I don't see the need to specify `endpoint` separately, it's more LOC and harder to follow imo